gcp/api/cursor.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""OSV API server cursor implementation."""
+
+import base64
+from enum import Enum
+from typing import Self
+import typing
+
+from google.cloud import ndb
+import google.cloud.ndb.exceptions as ndb_exceptions
+
+_FIRST_PAGE_TOKEN = base64.urlsafe_b64encode(b'FIRST_PAGE_TOKEN').decode()
+# Use ':' as the separator as it doesn't appear in urlsafe_b64encode
+# (which is what is used for both _FIRST_PAGE_TOKEN, and ndb.Cursor.urlsafe())
+_METADATA_SEPARATOR = ':'
+
+
+class _QueryCursorState(Enum):
```

Comment:
nit: Since it's the only thing that uses it , this could possibly go inside of `QueryCursor`:
```python
class QueryCursor:
  class _State(Enum):
    ENDED = 0
    # ...
    
  _cursor_state: _State = _State.ENDED

  @property
  def ended(self) -> bool:
    return self._cursor_state == QueryCursor._State.ENDED
````

That said, I don't think I have a preference either way.

Reply:
I think it's just a bit clearer to be private but at the top level to avoid having to write the class name repeatedly.